File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-go/internal/agente/acaogate.go
package agente

// AcaoProposta e uma acao que o agente quer executar.
//
// Adaptacao: no original, Executar e uma funcao/lambda sem argumentos
// embutida no objeto literal da acao proposta. Aqui vira um func() string
// dentro da struct — mesmo formato observavel (so roda se nao houver gate).
type AcaoProposta struct {
	Tipo            string
	RequerAprovacao bool
	Executar        func() string
}

// ResultadoAcao e o retorno de ExecutarOuGatear: ou foi "executada" (com
// Resultado preenchido), ou ficou "aguardando_aprovacao" (com Mensagem
// preenchida) — nunca os dois ao mesmo tempo, igual ao original.
type ResultadoAcao struct {
	Status    string
	Resultado string
	Mensagem  string
}

// ExecutarOuGatear e o Approval Gate: nem toda acao proposta executa direto.
func ExecutarOuGatear(acaoProposta AcaoProposta) ResultadoAcao {
	if acaoProposta.RequerAprovacao {
		return ResultadoAcao{
			Status: "aguardando_aprovacao",
			Mensagem: "Ação \"" + acaoProposta.Tipo + "\" NÃO executada. Aguardando Approval Gate " +
				"(Módulo 1.3, Pergunta 2: erro caro e irreversível).",
		}
	}
	return ResultadoAcao{
		Status:    "executada",
		Resultado: acaoProposta.Executar(),
	}
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-02-single-agent/agent-components-demo-java/src/main/java/com/trialforge/agentcomponents/AcaoGate.java
package com.trialforge.agentcomponents;

import java.util.function.Supplier;

/**
 * 4 e 5. ACAO + APPROVAL GATE — nem toda acao executa direto.
 *
 * Porte 1:1 de agent-components-demo.js / agent_components_demo.py (secao 4/5).
 */
public final class AcaoGate {

    private AcaoGate() {
    }

    /**
     * Adaptacao: no original, {@code executar} e uma funcao/lambda sem argumentos
     * embutida no objeto literal da acao proposta. Aqui vira um {@link Supplier}
     * dentro de um record — mesmo formato observavel (so roda se nao houver gate).
     */
    public record AcaoProposta(String tipo, boolean requerAprovacao, Supplier<String> executar) {
    }

    public record ResultadoAcao(String status, String resultado, String mensagem) {
        static ResultadoAcao executada(String resultado) {
            return new ResultadoAcao("executada", resultado, null);
        }

        static ResultadoAcao aguardandoAprovacao(String mensagem) {
            return new ResultadoAcao("aguardando_aprovacao", null, mensagem);
        }
    }

    public static ResultadoAcao executarOuGatear(AcaoProposta acaoProposta) {
        if (acaoProposta.requerAprovacao()) {
            return ResultadoAcao.aguardandoAprovacao(
                    "Ação \"" + acaoProposta.tipo() + "\" NÃO executada. Aguardando Approval Gate "
                            + "(Módulo 1.3, Pergunta 2: erro caro e irreversível)."
            );
        }
        return ResultadoAcao.executada(acaoProposta.executar().get());
    }
}
